include/substrait-mlir/Dialect/Substrait/IR/SubstraitTypes.td

```diff
@@ -180,14 +180,32 @@ def Substrait_AtomicTypes {
 /// One of the currently supported atomic types.
 def Substrait_AtomicType :
   AnyTypeOf<Substrait_SimpleTypes.types#Substrait_ParametrizedTypes.types>;
+  
+def Substrait_ListType : Substrait_Type<"List", "list"> {
+  let summary = "Substrait list type";
+  let description = [{
+    This type represents a list type.
+  }];
+  let parameters = (ins Substrait_AtomicType:$type);
+  let assemblyFormat = [{ `<` $type `>` }];
+  let genVerifyDecl = 1;
```

Comment:
I think it is possible and most certainly preferrable to verify this using tablegen (more). Take a look at `TupleOf` in `CommonTypeConstraints.td` upstream. (`NestedTupleOf` will then be helpful as inspiration for nested Substrait structs, lists, and maps...). One problem here is that we don't see in this file which types can be used, plus we now have to maintain *two* lists in two different files, which risk to run out of sync at some point.